### ISSUE TITLE
Fix false positives for final newlines in max-empty-lines

### ISSUE THREAD

**Issue by silverwind** [stylelint/stylelint#3722]
Since v9.6.0, `max-empty-lines` of value `0` reports the final newline in a file as an error. I can confirm this behaviour was not there in v9.5.0.

```` bash
$ echo "a {}" | npx stylelint

<input css 1>
 2:1  ✖  Expected no more than 0 empty lines   max-empty-lines
````

`.stylelintrc`:
```yaml
rules:
  max-empty-lines: 0
```

**Comment by jeddy3** [comment]
@silverwind Thanks for the report and for using the template.

This looks to be a regression. I'll label it up as a bug and help wanted.

**Comment by sswebcoder** [comment]
I think it's normal behaviour, because you using "echo" program, but "echo" just adding line-break after.

![image](https://user-images.githubusercontent.com/22288830/47682428-698e7800-dbdd-11e8-808d-df967de4ea6a.png)

As result we have two lines instead one.

**Comment by sswebcoder** [comment]
In text editor it looks like this:

![image](https://user-images.githubusercontent.com/22288830/47682575-e7eb1a00-dbdd-11e8-881c-36cb7953d5cc.png)

**Comment by silverwind** [comment]
Yes, I'm aware of that echo behaviour. I add final newlines to all my files for concatenation purposes and I have this enforced in my `.editorconfig`. It's still a bug in stylelint as per my definition, a empty line is `\n\n`, not `\n`.

**Comment by sswebcoder** [comment]
Thanks for explanation. I got it. I will try to debug.

**Comment by sswebcoder** [comment]
^[\s\t]*\r?\n

**Comment by GorwinJororis** [comment]
@silverwind @jeddy3 Well, I understand why you are adding newline for end of file beside wanting no empty lines between begin and end. But I do not share same thoughts as yours about what is empty line, what is not. 

Empty line is `\n\n` between begin and end of file like you said. But for begin or end line it is just one `\n`. You can see that in @sswebcoder 's screenshot. So we should consider this one `\n` is one empty line. But of course there can be option added to `max-empty-lines` which can allow us to use newline character in end of file.

**Comment by silverwind** [comment]
> But for begin or end line it is just one `\n`

I disagree. If you see a "empty" line or not depends on how your editor renders it. For example, `vim` in its default configuration does not render the final newline as a visual "empty" line:

````
$ cat file | hexdump -C
00000000  63 6f 6e 74 65 6e 74 0a                           |content.|
00000008
````
Yet if I open it:
````
$ vim file
````

````
content
~
````

Same goes for `cat` and other unix tools which need these empty lines to not mangle command output and prompt. So the fact that some editors display a literal empty line is a implementation detail. I think something like `\r?\n\r?\n` should be all that's needed.

**Comment by GorwinJororis** [comment]
Well, I disagree with both of you. 

You're telling me what unix systems were doing back then and are doing right now like they're one mighty knows everything rulers whose are making rules.

In seventies, there were good reasons to had rule like that: _"Every line of ALL human-readable text files must end with a newline character including last one"_

Reasons for that, we can say _"avoid extra processing when displaying or joining text files"_ or _"avoid treating text files differently to files containing other kinds of data (eg raw binary data which isn't human-readable)"_

Because of this convention, many tools from that era expect the ending newline, including text editors, diffing tools, and other text processing tools. Mac OS X was built on BSD Unix, and Linux was developed to be Unix-compatible, **so both operating systems have inherited the same convention, behaviour and tools.**

Windows wasn't developed to be Unix-compatible, so it doesn't have the same convention, and most **Windows software will deal just fine with no trailing newline**.

**As you can see maybe it made sense in 1971, but in our present, it is just convention and mostly maintaining compatibility with existing tools.**

Also _empty line aspect_ is not a computer thing. It is **term for humans**. So you **cannot rely on** what linux is saying about last newline character or even windows. 


![](https://user-images.githubusercontent.com/22288830/47682575-e7eb1a00-dbdd-11e8-881c-36cb7953d5cc.png)

For our eyes, **it is empty line** whatever you like it or not. **When you put newline character at end of line you are starting writing new line but never do.**

> @silverwind current implementation is correct? Can we close issue?

Also it is not cool saying this like I have nothing to say about it and asking to close the issue.

**Comment by sswebcoder** [comment]
What is empty line?
Escape symbol \n - is mean what after that starting new line.
But new line maybe be or not to be.
Empty line should has start and end, therefore it looks like a {} \n\n
there new line has start(first \n) and has the end(next \n).

Sorry for my english.

**Comment by sswebcoder** [comment]
Same file in different editors.

![image](https://user-images.githubusercontent.com/22288830/48376851-3075fd80-e6dd-11e8-804b-fe0f1714a2c5.png)

Two lines in VS Code and one in vim.

**Comment by silverwind** [comment]
This was fixed in https://github.com/stylelint/stylelint/pull/3785, thanks.
